cgen: add an error counter

amscanne opened the pull request:
For kfunc call errors, add a dedicated counter that can be read by the daemon in the future. This counter is exposed implicitly as a named counter object ("errors") via `nft`.

Fixes #105

qdeslandes reviewed (changes requested):
Thanks for doing this! 

Regarding `nft` support for counters, it's a bit of a pain at the moment as it diverged a lot from what `bfcli` can do. Also, if you send back the error counter, `nftables` will have no idea what to do with it, as that doesn't exist in `nft` world. 

To check if your counter is updated, the easiest solution would be to use `bpftool map dump id XXX` to dump the counters map (`bf_cmap_xxx`).

`bfcli` can't yet fetch the counters, but @ryantimwilson is working on it. In the meantime, I would suggest to go the `bpftool` way, as neither `iptables` nor `nftables` has any concept of error counter.

amscanne commented on src/bpfilter/cgen/cgen.h:
```diff
@@ -146,3 +149,29 @@ void bf_cgen_dump(const struct bf_cgen *cgen, prefix_t *prefix);
+/**
+ * Get packets and bytes counter for the policy.
+ *
+ * See `bf_cgen_get_counter` for an explanation.
```
Done

amscanne commented on src/bpfilter/cgen/cgen.h:
```diff
@@ -146,3 +149,29 @@ void bf_cgen_dump(const struct bf_cgen *cgen, prefix_t *prefix);
+ * @param cgen Codegen to get the counter for. Can't be NULL.
+ * @param counter Counter structure to fill with the counter values. Can't be
+ *        NULL.
+ * @return 0 on success, or a negative errno  value on failure.
```
Done

amscanne commented on src/bpfilter/cgen/cgen.c:
```diff
@@ -190,14 +190,37 @@ int bf_cgen_get_counter(const struct bf_cgen *cgen, uint32_t counter_idx,
+                                  bf_list_size(&cgen->chain->rules)+1,
+                                  counter);
+}
+
```
There are some other consequences for this, and I don't *love* abusing the enum as a index also, but I don't have a strong preference. PTAL, and see if this makes sense with the ipt changes to support these semantics (since it will need to change rather than relying on the size of the rules being used to access the policy).

qdeslandes replied:
It's not ideal, I agree, but I feel the intent is clearer. For `iptables`, the semantics used by `bpfilter` are different, so the bridging between `iptables` and `bpfilter` is expected (same for `nftables`).

amscanne replied:
I've done it in this way.

amscanne commented:
> Regarding nft support for counters, it's a bit of a pain at the moment as it diverged a lot from what bfcli can do. Also, if 
> you send back the error counter, nftables will have no idea what to do with it, as that doesn't exist in nft world.

I've gone this route and dropped the NFT code (since I think it opens a can of worms there). Do you know of a good way to force failures on these paths and test? Everything seems to be the way I expect it, but I'd like to test with real faults on these paths.

amscanne commented on src/bpfilter/cgen/program.c:
```diff
@@ -936,8 +936,9 @@ int bf_program_generate(struct bf_program *program)
-    // Add 1 to the number of counters for the policy counter.
-    program->num_counters = bf_list_size(&chain->rules) + 1;
+    // Add 1 to the number of counters for the policy counter, and 1
+    // for the first reserved error slot.
```
Done

qdeslandes commented on src/bpfilter/cgen/cgen.h:
```diff
@@ -127,12 +127,22 @@ int bf_cgen_unload(struct bf_cgen *cgen);
+/**
+ * @enum bf_counter_type
+ *
+ * Special counter types for @ref bf_cgen_get_counter.
```
`@ref bf_cgen_get_counter.` -> `@ref bf_cgen_get_counter .` otherwise Doxygen assumes the symbol name is `bf_cgen_get_counter.`

amscanne replied:
Done.

qdeslandes commented on src/bpfilter/cgen/cgen.h:
```diff
@@ -127,12 +127,22 @@ int bf_cgen_unload(struct bf_cgen *cgen);
- * counter in the map than the number of rules. This last counter (the last in
- * the map) is dedicated to the policy.
+ * Counters are referenced by their index in the counters map or the enum
+ * values defined by @ref bf_counter_type.
```
Same

amscanne replied:
Done

qdeslandes commented:
> I've gone this route and dropped the NFT code (since I think it opens a can of worms there). Do you know of a good way to force failures on these paths and test? Everything seems to be the way I expect it, but I'd like to test with real faults on these paths.

Unfortunately, no, as we would need to trigger an error on a `kfunc`/BPF helper. A well crafted BPF program could probably hook to one of the `kfunc` and filter the calls (failing those coming from `bpfilter`).

You could change the conditional used to check for error, so the counters would be updated on success, that would be better than nothing. 

Furthermore, could you run `make fixstyle` to apply the same style as the whole codebase?

amscanne commented:
> > I've gone this route and dropped the NFT code (since I think it opens a can of worms there). Do you know of a good way to force failures on these paths and test? Everything seems to be the way I expect it, but I'd like to test with real faults on these paths.
> 
> Unfortunately, no, as we would need to trigger an error on a `kfunc`/BPF helper. A well crafted BPF program could probably hook to one of the `kfunc` and filter the calls (failing those coming from `bpfilter`).
> 
> You could change the conditional used to check for error, so the counters would be updated on success, that would be better than nothing.
> 
> Furthermore, could you run `make fixstyle` to apply the same style as the whole codebase?

I've fixed the style as suggested. I ran some manual tests with the same counter in the valid path, and it seemed fine.

I don't want to complicate this specific pull request, but thinking out loud: did adding a dynamic layer of functions for fault injection? (E.g. dynamically add a function indirection where the builtins are all bounced through a trampoline, which could probabalistically or deterministically inject various faults.) Is such a feature generally useful? Maybe alongside being able to control the fail open/closed behavior for each chain? (With the fault injection basically giving you the ability to test that.)

qdeslandes commented:
> I don't want to complicate this specific pull request, but thinking out loud: did adding a dynamic layer of functions for fault injection? (E.g. dynamically add a function indirection where the builtins are all bounced through a trampoline, which could probabalistically or deterministically inject various faults.) Is such a feature generally useful? Maybe alongside being able to control the fail open/closed behavior for each chain? (With the fault injection basically giving you the ability to test that.)

That's a great idea, and it would definitely be useful for more than testing `kfunc` error handling (e.g. validate user-provided BPF bytecode). Please don't hesitate to give it a go if you're interested. I think most of the pieces are already available (map support, custom function generation).

qdeslandes commented:
LGTM. The CI is mad because of `clang-tidy`, if you rebase on `main` + add the following to `cgen.c` it will fix it:
```
diff --git a/src/bpfilter/cgen/stub.c b/src/bpfilter/cgen/stub.c
index 25bb6a3..9be6ce0 100644
--- a/src/bpfilter/cgen/stub.c
+++ b/src/bpfilter/cgen/stub.c
@@ -19,6 +19,7 @@
 #include <endian.h>
 #include <stddef.h>
 
+#include "bpfilter/cgen/fixup.h"
 #include "bpfilter/cgen/jmp.h"
 #include "bpfilter/cgen/printer.h"
 #include "bpfilter/cgen/program.h"
 ```

qdeslandes commented:
Done! Thanks @amscanne for doing this!
